src/wallet/wallet_rpc_server_mms_defs.h

```diff
@@ -0,0 +1,732 @@
+// Copyright (c) 2019, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers
+
+#pragma once
+#include "cryptonote_protocol/cryptonote_protocol_defs.h"
+#include "cryptonote_basic/cryptonote_basic.h"
+// #include "wallet/message_store.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "wallet.rpc"
+
+// When making *any* change here, bump versions in "wallet_rpc_server_commands_defs.h"
+// Check WALLET_RPC_VERSION_MAJOR and WALLET_RPC_VERSION_MINOR there early in the file
+
+
+// The following public methods of "message_store.h" don't have any equivalent here,
+// in an attempt to avoid unnecessary bloat:
+// - get_sanitized_message_text: If you display message text, e.g. notes, "sanitize"
+// yourself as needed in accordance with the environment you are in
+// - stop: Internal use by "wallet2"; use its 'stop' method if necessary
+// - read_file, write_file: Internal use by "wallet2"
+// - get_signer_index_by_monero_address, get_signer_index_by_label: Should be quite
+// trivial to implement in a client, with the help of "get_all_signers"
+
+namespace tools
+{
+namespace wallet_rpc
+{
+
+  struct mms_message
```

Comment:
These are duplicating `struct`s in `message_store.h` entirely. Please merge their usage somehow.

Reply:
This duplication of structs is along the same lines as some similar duplications of `wallet2` structs in `wallet_rpc_server_commands_defs.h`. The structs in `message_store.h` are "binary", so to say, with strings that directly hold binary contents, and use of enumeration type values. This is straight-forwared and simple for `simplewallet` as a **direct** user of the message store.

The structs here are suitable for RPC purposes, with enum values translated to strings (which I found to be current "best practice" after a quite thourough survey of opinions in this matter), and with any binary content like hashes, addresses and tx data hexed.

In my opinion these differences in usage merit the rathe simple code to translate back-and-forth.

Reply:
> The structs in message_store.h are "binary", so to say, with strings that directly hold binary contents, and use of enumeration type values. This is straight-forwared and simple for simplewallet as a direct user of the message store.

The serialization will differ but the struct do not need to differ. There's a discussion thread about adding JSON support to `boost::serialization` for instance. I recall pushing for JSON in that initial review (maybe I am mistaken), because the format of `boost::serialization` isn't particular well documented for interoperability. Now, we have to add a second serialization "strategy" (epee and `boost::serialization`) and format (json and custom boost). And each strategy supports multiple formats.

> The structs here are suitable for RPC purposes, with enum values translated to strings (which I found to be current "best practice" after a quite thourough survey of opinions in this matter), and with any binary content like hashes, addresses and tx data hexed.

The structs can use multiple serialization techniques. The `boost::serialization` free functions are not overloading on the same name as the epee json function name.

Reply:
> The structs can use multiple serialization techniques. The boost::serialization free functions are not overloading on the same name as the epee json function name.

How would that look in detail? I only know both serializations as highly automatic things so far, especially epee's JSON serialization. Where and how would there be a chance to intervene and serialize an enum as a string?